### PR TITLE
[285] SQL statements

### LINKED ISSUE #285: Allow querying DataFrame directly in sql method
While I really like the idea of @rxin's recent #256 PR, he uses an (in my opinion) over-simplistic example of `ks.sql("select * from range(10) where id > 7")`. I believe that the ability to query actual Koalas DataFrames through SQL can prove really valuable to many users. However, when trying to use `ks.sql` with a Koalas DataFrame, the following exception occurs:

```python
kdf = ks.DataFrame({'A': [1,2,3]})
>>> ks.sql("select * from kdf")
...
org.apache.spark.sql.AnalysisException: Table or view not found: kdf; line 1 pos 14
...
```

This is not surprising to someone with PySpark knowledge who knows that `kdf` has to be registered as a temporary table before being able to use it with SparkSQL. Unfortunately, (as I understand it) the target group of the Koalas library should not be expected to be Spark experts. To get the above example working, the following workaround is needed, which requires the usage of the lower-level (Py)Spark API, thus somewhat defeating the purpose of Koalas.

```python
>>> from pyspark import SparkContext
>>> from pyspark.sql import SQLContext
>>> sc = SparkContext.getOrCreate()
>>> sql = SQLContext(sc)
>>> sql.registerDataFrameAsTable(kdf._sdf, "kdf")
>>> ks.sql("select * from kdf")
   __index_level_0__  A                                                         
0                  0  1
1                  1  2
2                  2  3
# Optionally clean-up by dropping the temporary table
>>> sql.dropTempTable("kdf")
```

Wouldn't it be much more convenient if this "temporary table magic" would instead be handled by Koalas behind the scenes or are there any design objections against such an approach?

### PR DESCRIPTION
**Pull request by thunterdb**
This PR builds on #291  and adds SQL statement with curly brace escaping in the query, as discussed in #285. In particular, it allows arbitrary mixing of Koalas, pandas dataframes and scalar types, see the docstring.

The doctests should give reasonable coverage.

Note: dates are not supported for now but can be added later.

Resolves #285

### CODE REVIEW

File: databricks/koalas/sql.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            if df_id not in self._temp_views:
+                sdf = var.to_spark()
+                # We should not capture extra index.
+                # TODO: document this
+                sdf = sdf[list(var.columns)]
+                sdf.createOrReplaceTempView(df_id)
+                self._temp_views[df_id] = sdf
+            return df_id
+        if isinstance(var, str):
+            return pymysql.escape_string(str)
```
Comment (HyukjinKwon) [inline databricks/koalas/sql.py]: Do we need this `pymysql` dependency only to escape string?

File: databricks/koalas/sql.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            # Simply propagate PySpark exceptions
+            s = ("Could not execute statement: normalized_statement='{}' "
+                 "temp_views={}".format(self._normalized_statement, self._temp_views))
+            raise ValidationError(s, e)
+        finally:
+            for v in self._temp_views:
+                self._session.catalog.dropTempView(v)
+        return DataFrame(sdf)
+
+    def _convert(self, key) -> Any:
```
Comment (rxin) [inline databricks/koalas/sql.py]: need doc

File: databricks/koalas/sql.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+
+class SQLProcessor(object):
+
+    def __init__(self, scope: Dict[str, Any], statement: str, session: SparkSession):
+        self._scope = scope
+        self._statement = statement
+        # All the temporary views created when executing this statement
+        self._temp_views = {}  # type: Dict[str, SDataFrame]
+        self._cached_vars = {}  # type: Dict[str, Any]
```
Comment (rxin) [inline databricks/koalas/sql.py]: doc

File: databricks/koalas/sql.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        self._normalized_statement = None  # type: Optional[str]
```
Comment (rxin) [inline databricks/koalas/sql.py]: doc on what normalized mean

File: requirements-dev.txt [hunk 5]
```diff
@@ -2,6 +2,7 @@
 pandas>=0.23
 pyarrow>=0.10
 numpy>=1.14
+pymysql
```
Comment (rxin) [inline requirements-dev.txt]: is it just one function that we need? If that's the case, can we get rid of this dependency?

File: databricks/koalas/sql.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            return str(var)
+        if isinstance(var, Series):
+            return self._convert_var(var.to_dataframe())
+        if isinstance(var, pd.DataFrame):
+            return self._convert_var(ks.DataFrame(var))
+        if isinstance(var, DataFrame):
+            df_id = "koalas_" + str(id(var))
+            if df_id not in self._temp_views:
+                sdf = var.to_spark()
+                # We should not capture extra index.
```
Comment (rxin) [inline databricks/koalas/sql.py]: i feel we need top keep the index, otherwise there is no way to query the index. Index is just a normal column in SQL.

**Comment by rxin** [comment]
I'm going to close this one. I've taken over the work since Tim is busy with other stuff. 

New PR in: https://github.com/databricks/koalas/pull/448

**Pull request closed without merging**